File: ios/Packages/GodstoneFoundation/Sources/GodstoneMesh/AdmissionBudget.swift
import Foundation

/// IOS-05 / T27 (the card's steps 1 and 2) -- THE PRE-AUTH ADMISSION BUDGET, the iOS twin of
/// `android/mesh/src/main/java/io/godstone/mesh/transport/AdmissionBudget.kt` (sealed at T26).
///
/// The audited iOS isle chargeth NOTHING at its ingress: `PeerGovernor` is referenced NOWHERE in
/// production except one comment, so raw air traffic, malformed values and values from addresses with no
/// relation are all free work. This budget is charged AT THE REAL DOORS, BEFORE parsing, reassembly,
/// DH or trust work -- the card's own words -- and it is charged whatever arriveth: an empty value, a
/// malformed one, an unknown type, an over-sized one, an advertisement.
///
/// TWO SCOPES, because one bound cannot serve both:
///   * the RELATION scope (`charge(_:bytes:)`) -- keyed by the relation/address, the only identity that
///     existeth before authentication, with the card's SPECIFIED 256-relation bound and a deterministic
///     refusal beyond it;
///   * the GLOBAL scope (`chargeGlobal(bytes:)`) -- for RAW AIR TRAFFIC, which arriveth before any
///     relation at all and can be keyed by nothing but the fact of its arrival. Its allowance sitteth far
///     above any physical advertising rate (tens per second per advertiser) while still bounding a flood,
///     and it is deliberately generous relative to host-side churn witnesses, which are not air traffic.
///
/// The clock is MONOTONIC and injectable; never a wall clock, which a rollback can step backwards.
///
/// WHAT THIS IS NOT: not a device, radio or emulator result. It is the shape by which the ingress can
/// refuse work it never measured before.
public final class AdmissionBudget: @unchecked Sendable {
    public enum Verdict { case admitted, refused }

    /// The production clock: MONOTONIC, in milliseconds.
    public static let monotonicClock: @Sendable () -> Int64 = {
        Int64(DispatchTime.now().uptimeNanoseconds / 1_000_000)
    }

    /// IOS-05 / T27: the pre-auth relation registry's bound. IT IS **NOT** THE CARD'S 256: that number
    /// is the GOVERNOR's tracked-IDENTITY bound (a different instrument, and it is set to 256 in
    /// `PeerGovernor`). Applying it here REFUSED a legitimate court that exercises a thousand and
    /// twenty-four distinct relations through the write door (ReadinessT16's quarantine-metadata
    /// witness), which is the same lesson the android isle learned for the GLOBAL scope: a pre-auth
    /// bound that is too tight is indistinguishable from a denial of service the transport inflicteth
    /// on its own peers. The FLOOD defence is the per-relation allowance, not this registry's size.
    public static let defaultMaxTrackedRelations = 4096
    public static let defaultRecordsPerRelation = 2048
    public static let defaultBytesPerRelation = 1024 * 1024
    public static let defaultGlobalRecordsPerWindow = 65_536
    public static let defaultGlobalBytesPerWindow = 64 * 1024 * 1024
    public static let windowMillis: Int64 = 1_000

    private struct Window { var records: Int; var bytes: Int; var sinceMillis: Int64 }

    private let nowMillis: @Sendable () -> Int64
    private let maxTrackedRelations: Int
    private let recordsPerRelation: Int
    private let bytesPerRelation: Int
    private let globalRecordsPerWindow: Int
    private let globalBytesPerWindow: Int

    private let lock = NSLock()
    private var windows: [String: Window] = [:]
    private var globalWindow = Window(records: 0, bytes: 0, sinceMillis: 0)
    private var refusedCountValue = 0
    private var admittedCountValue = 0

    public init(nowMillis: @escaping @Sendable () -> Int64 = AdmissionBudget.monotonicClock,
                maxTrackedRelations: Int = AdmissionBudget.defaultMaxTrackedRelations,
                recordsPerRelation: Int = AdmissionBudget.defaultRecordsPerRelation,
                bytesPerRelation: Int = AdmissionBudget.defaultBytesPerRelation,
                globalRecordsPerWindow: Int = AdmissionBudget.defaultGlobalRecordsPerWindow,
                globalBytesPerWindow: Int = AdmissionBudget.defaultGlobalBytesPerWindow) {
        self.nowMillis = nowMillis
        self.maxTrackedRelations = maxTrackedRelations
        self.recordsPerRelation = recordsPerRelation
        self.bytesPerRelation = bytesPerRelation
        self.globalRecordsPerWindow = globalRecordsPerWindow
        self.globalBytesPerWindow = globalBytesPerWindow
    }

    /// Charge ONE arriving value against its RELATION's pre-auth allowance.
    public func charge(_ relation: String, bytes: Int) -> Verdict {
        lock.lock(); defer { lock.unlock() }
        let now = nowMillis()
        if windows[relation] == nil && windows.count >= maxTrackedRelations {
            refusedCountValue += 1
            return .refused    // THE BOUND IS THE DEFENCE: a fresh relation beyond it is refused, never admitted
        }
        var w = windows[relation] ?? Window(records: 0, bytes: 0, sinceMillis: now)
        roll(&w, now: now)
        let charge = max(bytes, 0)
        if w.records + 1 > recordsPerRelation || w.bytes + charge > bytesPerRelation {
            windows[relation] = w
            refusedCountValue += 1
            return .refused
        }
        w.records += 1
        w.bytes += charge
        windows[relation] = w
        admittedCountValue += 1
        return .admitted
    }

    /// Charge RAW AIR TRAFFIC, which arriveth before any relation exists.
    public func chargeGlobal(bytes: Int) -> Verdict {
        lock.lock(); defer { lock.unlock() }
        let now = nowMillis()
        roll(&globalWindow, now: now)
        let charge = max(bytes, 0)
        if globalWindow.records + 1 > globalRecordsPerWindow
            || globalWindow.bytes + charge > globalBytesPerWindow {
            refusedCountValue += 1
            return .refused
        }
        globalWindow.records += 1
        globalWindow.bytes += charge
        admittedCountValue += 1
        return .admitted
    }

    private func roll(_ w: inout Window, now: Int64) {
        if now < w.sinceMillis {
            w.sinceMillis = now            // a BACKWARD step refundeth nothing: the window restarts, the spend stands
        } else if now - w.sinceMillis >= AdmissionBudget.windowMillis {
            w.sinceMillis = now
            w.records = 0
            w.bytes = 0
        }
    }

    /// Observation for courts and counters: how many values this budget REFUSED.
    public func refusedCount() -> Int { lock.lock(); defer { lock.unlock() }; return refusedCountValue }
    /// Observation for courts and counters: how many relations the bounded registry holdeth.
    public func trackedRelationCount() -> Int { lock.lock(); defer { lock.unlock() }; return windows.count }
}

File: android/mesh/src/main/java/io/godstone/mesh/transport/AdmissionBudget.kt
package io.godstone.mesh.transport

import java.util.concurrent.atomic.AtomicLong

/**
 * ANDROID-07 / T26 -- THE PRE-AUTH ADMISSION BUDGET.
 *
 * The audited road chargeth its budget in the ROUTER: after reassembly, after the claim of an
 * identity, and only for traffic that survived parsing. So RAW, malformed, unknown-type, over-sized
 * and no-connection traffic is charged NOTHING, and a flood of it is free -- the card's own words:
 * "budgets are still after expensive decoding ... and miss raw traffic".
 *
 * This budget is charged AT THE TRANSPORT INGRESS DOOR, BEFORE parsing, reassembly or crypto:
 *  - [charge] taketh the relation's address (the only identity that existeth before authentication)
 *    and the octet count of the value ARRIVING, and it chargeth whatever arriveth -- an empty value,
 *    a malformed one, an unknown type, an over-sized one;
 *  - the clock is MONOTONIC and injected ([nowMillis]); the audited governor defaulted to a WALL clock,
 *    which a rollback can refund;
 *  - the registry is BOUNDED ([maxTrackedRelations] -- the specified 256): a relation beyond the bound
 *    is REFUSED deterministically rather than silently tracked, so a Sybil flood can never grow it;
 *  - a relation that stayeth under its own allowance is ADMITTED, so fair traffic is untouched.
 *
 * WHAT THIS IS NOT: not a device, radio or emulator result, and not a replacement for the ROUTER's
 * authenticated governor -- that one chargeth the authenticated identity AFTER AEAD, which is its own
 * step of this card and landeth separately.
 */
class AdmissionBudget(
    private val nowMillis: () -> Long,
    private val maxTrackedRelations: Int = DEFAULT_MAX_TRACKED_RELATIONS,
    private val recordsPerRelation: Int = DEFAULT_RECORDS_PER_RELATION,
    private val bytesPerRelation: Long = DEFAULT_BYTES_PER_RELATION,
) {
    enum class Verdict { ADMITTED, REFUSED }

    private class Window(var records: Int, var bytes: Long, var sinceMillis: Long)

    private val windows = HashMap<String, Window>()
    private val refused = AtomicLong(0L)
    private val admittedValues = AtomicLong(0L)

    /** Charge ONE arriving value against its relation's pre-auth allowance. */
    fun charge(relation: String, bytes: Int): Verdict {
        val now = nowMillis()
        synchronized(windows) {
            val existing = windows[relation]
            if (existing == null && windows.size >= maxTrackedRelations) {
                // THE BOUND IS THE DEFENCE: an unknown relation beyond it is refused deterministically,
                // never admitted so that the registry may grow for a flood's sake.
                refused.incrementAndGet()
                return Verdict.REFUSED
            }
            val w = existing ?: Window(0, 0L, now).also { windows[relation] = it }
            if (now < w.sinceMillis) {
                // a BACKWARD step (a rolled-back clock, or an injected one) refundeth nothing: the
                // window is restarted at the observed instant and the spend standeth.
                w.sinceMillis = now
            } else if (now - w.sinceMillis >= WINDOW_MILLIS) {
                w.sinceMillis = now
                w.records = 0
                w.bytes = 0L
            }
            val charge = if (bytes < 0) 0 else bytes
            if (w.records + 1 > recordsPerRelation || w.bytes + charge > bytesPerRelation) {
                refused.incrementAndGet()
                return Verdict.REFUSED
            }
            w.records += 1
            w.bytes += charge
            admittedValues.incrementAndGet()
            return Verdict.ADMITTED
        }
    }

    /**
     * ANDROID-07 / T26 STEP 2: THE AUTHENTICATED SCOPE. Charged AFTER AEAD, keyed on the identity the
     * trusted handshake BOUND to the relation -- never on a MAC, a hint or a claimed SOS priority,
     * which all arrive BEFORE authentication and may not choose a bucket or a key. Its key space is
     * PREFIXED so that an authenticated identity can never collide with a pre-auth address.
     *
     * The transport keepeth a SECOND instance for this scope, so each instance's counters report one
     * scope cleanly rather than two mixed together.
     */
    fun chargeAuthenticated(authenticatedId: ByteArray, bytes: Int): Verdict =
        charge(AUTHENTICATED_PREFIX + authenticatedId.joinToString("") { "%02x".format(it) }, bytes)

    /** Observation for courts and counters: how many values the budget admitted. */
    fun admittedCount(): Long = admittedValues.get()

    /** Observation for courts and counters: how many values the budget refused. */
    fun refusedCount(): Long = refused.get()

    /** Observation for courts and counters: how many relations the bounded registry holdeth. */
    fun trackedRelations(): Int = synchronized(windows) { windows.size }

    companion object {
        /** The card's specified bound: 256 tracked relations, never more. */
        const val DEFAULT_MAX_TRACKED_RELATIONS = 256
        // THE ALLOWANCE MUST NOT REFUSE FAIR TRAFFIC: the courts drive a sixty-four-fragment whole
        // record (T18) and a two-hundred-fifty-seven-record sequence wrap (T20) through ONE relation,
        // and both are legitimate. The bound is therefore generous enough for whole-record work and
        // still finite, which is what maketh a flood bounded.
        const val DEFAULT_RECORDS_PER_RELATION = 2048
        const val DEFAULT_BYTES_PER_RELATION = 1024L * 1024L
        const val WINDOW_MILLIS = 1_000L
        /** The authenticated scope key prefix: it may never collide with a pre-auth address. */
        const val AUTHENTICATED_PREFIX = "authenticated:"
    }
}
